### PR TITLE
Clients REST API generation

### PR DESCRIPTION
**Pull request by adamjez**
**_Do not merge: Work In Progress_**

This PR is introducing two new parameters to Open API specification generated with Swashbuckle: `x-dotvvm-known-type` and `x-dotvvm-name`. These parameters provide required information for NSwag when generating **TypeScript** and **C#** clients. 

* Specifically `x-dotvvm-known-type` is present for types that exist in both **REST API project** and **DotVVM project** so NSwag doesn't generate DTOs for these types. Generic types get special treatment.
* `x-dotvvm-name` specifies *names* for known types.

Additionally, there is new attribute `HttpMethod` to annotate client C# methods with information about used http method.

Currently, It's all implemented only for AspNetCore. I plan to port it when the following issues are resolved.

Unresolved issues:
* Known types are provided in `DotvvmApiKnownTypesOptions` as `Type[]`. Filters get this options object via `IOptions` so users can add their own types. DotVVM types are _hardcoded_ into the options object, because in method `EnableDotvvmIntegration` I don't have access to service provider.
* `DotVVM.Framework.Api.Swashbuckle.AspNetCore` project is only referencing `DotVVM.Core` so logic for determining property name when serializing in `DotvvmViewModel` can't be shared and therefore is copied inside API project. 

Resolves https://github.com/riganti/dotvvm/issues/554

### CODE REVIEW

File: src/DotVVM.Framework/Compilation/BindingRequiredResourceVisitor.cs [hunk 1]
```diff
@@ -23,61 +23,57 @@ public BindingRequiredResourceVisitor(ControlResolverMetadata requiredResourceCo
...
 
         public override void VisitPropertyBinding(ResolvedPropertyBinding propertyBinding)
         {
-            var r = propertyBinding.Binding.Binding.GetProperty<RequiredRuntimeResourcesBindingProperty>(ErrorHandlingMode.ReturnNull);
-            if (r != null)
-                requiredResources = requiredResources.Union(r.Resources);
+            var resourceProperty = propertyBinding.Binding.Binding.GetProperty<RequiredRuntimeResourcesBindingProperty>(ErrorHandlingMode.ReturnNull);
+            if (resourceProperty != null)
+            {
+                requiredResources = requiredResources.Union(resourceProperty.Resources);
```
Comment (exyi) [inline src/DotVVM.Framework/Compilation/BindingRequiredResourceVisitor.cs]: I'm sorry, but this change does not make it more readable, I actually struggle to visually recognize. I'd argue that single-letter name like `r` is completely reasonable in this context.

```
resourceProperty and requiredResources
```

Comment (adamjez) [reply src/DotVVM.Framework/Compilation/BindingRequiredResourceVisitor.cs]: I admit that in this situation, more descriptive naming of variable doesn't add too much, but in the future you can add more statements to the function and from single-letter name like `r` you will not be able to determine what the variable represents. If you struggle with recognizing `resourceProperty` and `requiredResources` I can change it to `property` if you find to more recognizable.

Comment (exyi) [reply src/DotVVM.Framework/Compilation/BindingRequiredResourceVisitor.cs]: That would be better for visual recognision but even less descriptive than `r`. In this context it's totally unimportant that it's a binding property, the important fact is that it contains the  resource requirements of this binding.

Comment (adamjez) [reply src/DotVVM.Framework/Compilation/BindingRequiredResourceVisitor.cs]: ok, so in this situation I'd choose `requiredResourceProperty`. @djanosik what do you think?

**Comment by adamjez** [comment]
I ported client generation to Owin projects. Unfortunatly I can't run DI tests on this PR because branch is located in my own repo. I ran all tests locally and one test is failing: `JavascriptCompilation_Api_DateParameter` but I'm not aware about any changes in this area.

About the issues mentioned in the main post. 
* I moved the logic for determining property name in ViewModel into separate class which is in **DotVVM.Core** project.
* I let DotVVM known types _hardcoded_ in the `DotvvmApiOptions` because I didn't come with better solution.
* In Swashbuckle integration for Owin I cannot use `IOptions` wrapper because swashbuckle filters are not resolved via DI. User can add its own known types via `Action<DotvvmApiOptions>` delegate which is parameter in `EnableDotvvmIntegration` configuration method. ThirdParty libs can add theirs known types with extensions method on `DotvvmApiOptions`.

**Pull request merged**